### PR TITLE
Improvements to workflows (MEN-2960, MEN-2993, MEN-3011, MEN-3012)

### PR DESCRIPTION
**Pull request by tranchitella**
**MEN-3012**: Improve error handling of HTTP tasks:

- Workflow definition support nows an optional list of response status codes to be considered valid. If the list is populated with at least one item, the HTTP status code of the response is checked to determine if the task was successful or not.
- If a workflow task fails, the workflow is stopped and the status is set to StatusFailure.

**MEN-2993**: Add "cli" task type to run command line-based tasks:
- We introduce a new "cli" task type, which allows to run command line commands and capture standard output and standard error. The exit code determines if the task is successful or not: exit code 0 is successful, otherwise, the task failed and the workflow execution is stopped.
- This new task type supports an optional timeout to limit the execution time of the command. The default execution time is 4 hours.

**MEN-2960**: Improve test code coverage:
- Add unit tests for store/mongo, using external mongo instance
- Refactored package main to simplify unit testing
- Minor cosmetic changes

### CODE REVIEW

**Comment by tranchitella** [comment]
@lluiscampos can you please review my approach to run the unit tests that need a mongodb instance?

File: api/http/workflow_test.go [hunk 1]
```diff
@@ -105,6 +140,180 @@ func TestWorkflowFoundAndLaunchedWithParameters(t *testing.T) {
...
 	assert.Equal(t, 1, len(dataStore.Jobs))
+
+	var response map[string]string
+	body := w.Body.Bytes()
+	err = json.Unmarshal(body, &response)
+	value, ok := response["id"]
+	assert.True(t, ok)
+
+	w = httptest.NewRecorder()
+	req, _ = http.NewRequest("GET", fmt.Sprintf("/api/v1/workflow/test/%s", value), nil)
```
Comment (merlin-northern) [inline api/http/workflow_test.go]: we are including fmt only for this one call, cant we just concatenate the strings here?

**Comment by tranchitella** [comment]
@merlin-northern I don't us dbtest because it requires me to have a "mongod" locally, which is an anti-pattern. With this repo I'm experimenting using an external mongodb (in this specific case, a service in gitlab) in order to run the tests independently from a local mongod binary.

I believe it is a better approach, it allows me to run the tests against different versions of mongo (or even different configurations of mongo, e.g., replicasets). What's your opinion about it?

**Comment by merlin-northern** [comment]
> @merlin-northern I don't us dbtest because it requires me to have a "mongod" locally, which is an anti-pattern. With this repo I'm experimenting using an external mongodb (in this specific case, a service in gitlab) in order to run the tests independently from a local mongod binary.
> 
> I believe it is a better approach, it allows me to run the tests against different versions of mongo (or even different configurations of mongo, e.g., replicasets). What's your opinion about it?

in all services we are using dbtest so I am not sure it is a good place to change it now.
with external service it means that when I want to run tests locally I have to have access to external db, which I do not like. I think tests should be self-contained. Otherwise the use of external db has some advantages. I would add github.com/mendersoftware/go-lib-micro/mongo/dbtest_external or something like that, so we follow the same interface and can use it when appropriate.

**Comment by tranchitella** [comment]
> in all services we are using dbtest so I am not sure it is a good place to change it now.
> with external service it means that when I want to run tests locally I have to have access to external db, which I do not like. I think tests should be self-contained. Otherwise the use of external db has some advantages. I would add github.com/mendersoftware/go-lib-micro/mongo/dbtest_external or something like that, so we follow the same interface and can use it when appropriate.

@merlin-northern you can run tests locally in the very same way you run tests locally with the other services. We are talking about "long" tests, the ones that run without the `-short` flag. `go-lib-micro/mongo/dbtest` requires you to have a `mongod` binary, while `workflows` use an external mongod (either locally installed, docker-compose based, or anything else). As I've said, this is a new service and I believe it allows us to experiment, that's why I'm asking for feedback. If this turns out to be a good pattern we can provide it as part of the `go-lib-micro` package, true.

Answering your second question, we are talking about datastore_mongo tests, where I really need the mongodb instance. All the other unit tests use a mocked data_store and don't need a real mongo instance.

File: api/http/workflow_test.go (hunk 1 above)
Comment (tranchitella) [reply api/http/workflow_test.go]: @merlin-northern right, I've removed the usage of `fmt` and improved the test code using the constant used to define the URLs in the router.

File: api/http/workflow_test.go [hunk 2]
```diff
@@ -40,7 +41,7 @@ func TestWorkflowNotFound(t *testing.T) {
 	req, _ := http.NewRequest("POST", "/api/v1/workflow/test", strings.NewReader(payload))
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/workflow_test.go]: @merlin-northern right, I've removed the usage of strings and improved the test code using the constant used to define the URLs in the router.

**Comment by mchalski** [comment]
@tranchitella I have some remarks, round 1. I'm afraid I'll have to get back to it once again after some initial discussion.

generally regarding the PR:
- let's keep them much **smaller** in the future, please. it's a lot to digest, and hard to read
- please avoid one thing I mentioned earlier - **commits that fix previous commits** in the same PR. case in point is the datastore mock - first it's implemented manually, then a couple commits later migrated to `testify`. when reading commits in sequence, it's surprising and doesn't help readability. in such cases we just `git rebase -i` the fix back into its proper commit
- I see a **merge commit** at the very end - doesn't it mean you had a stale branch with respect to master? then, before submitting the PR, rebase on master first, and you'll get a clean commit log

now about the specifics, starting from the easier stuff:
- I'd second Peter's remarks on **test cases.** we typically use them as it's easy to add more and more corner cases. I won't CR on this, because your method is also valid, but in the future, please try incorporating them (at least a bit more - even if you do method-per-scenario, they can be useful, for instance for bad request testing; then it's `TestFooBadRequest` and test cases can capture multiple combinations of failing params)
- btw good to see you used `mock` for the db mock in the end
- I need more clarifications on dropping **dbtest** though; we're still using a real database, right? only now it's parametrized with a config value? is it all there is to it?
- also need more clarifications on **workflow include/exclude** - why is it needed? it seems like a smell, wouldn't it be enough to just give the worker the workflows it needs? I'm sure Im not understanding this
- the bigger thing is **cli task in place of async task**; I'm not excited about this because this is not how it was described in the task/epic. I know that image generation can be generalized as a sequence of cli commands, but I'm afraid at some point we'll just want a freeform async task where you can do whatever - call clis, other services, etc. I won't CR this but please let's avoid such surprises. sometimes we do change the task specs just-in-time when they are obviously wrong, but this was not strictly necessary here
- nitpicks on the **TaskResult struct/API** - I feel we can do better here. instead of optional field-per-task-type, maybe just a `result` that has subtypes? in general, some polymorphic behavior would fit here I think

**Comment by tranchitella** [comment]
@mchalski 

I know, this PR is a complex one: it is the result of several weeks of work. In fact I've had three different PRs, but as nobody was available to review each of them was based on the previous one and this is the result:

https://github.com/mendersoftware/workflows/pull/10
https://github.com/mendersoftware/workflows/pull/11

> * let's keep them much **smaller** in the future, please. it's a lot to digest, and hard to read

Special case here, nobody was around for two weeks and a lot of work was needed on workflows. I know this is not ideal, sorry.

> * please avoid one thing I mentioned earlier - **commits that fix previous commits** in the same PR. case in point is the datastore mock - first it's implemented manually, then a couple commits later migrated to `testify`. when reading commits in sequence, it's surprising and doesn't help readability. in such cases we just `git rebase -i` the fix back into its proper commit

See above: this PR is the result of three different PRs. I can rebase if it helps you, but in fact reviewing the code as it is now, without reviewing the single commits, is almost equivalent.

> * I see a **merge commit** at the very end - doesn't it mean you had a stale branch with respect to master? then, before submitting the PR, rebase on master first, and you'll get a clean commit log

A PR got merged to master, I could have rebased, you are right.

> * I'd second Peter's remarks on **test cases.** we typically use them as it's easy to add more and more corner cases. I won't CR on this, because your method is also valid, but in the future, please try incorporating them (at least a bit more - even if you do method-per-scenario, they can be useful, for instance for bad request testing; then it's `TestFooBadRequest` and test cases can capture multiple combinations of failing params)

I've used the TestCases approach, I believe, where applicable.

> * btw good to see you used `mock` for the db mock in the end

Yes, I wasn't aware of the usage of the mock library.

> * I need more clarifications on dropping **dbtest** though; we're still using a real database, right? only now it's parametrized with a config value? is it all there is to it?

The main difference is that dbtest is starting and stopping the `mongod` binary. My pattern is using an external database (parametrized via env variable), which makes it simple and easy to test the package against different versions (or different setups, e.g., replicasets) of mongodb. This is the way I usually write tests that need external resources, and makes testing locally simple and easy using docker-compose to bring up a mongo instance.

> * also need more clarifications on **workflow include/exclude** - why is it needed? it seems like a smell, wouldn't it be enough to just give the worker the workflows it needs? I'm sure Im not understanding this

Well, it limits the worker to act on a specific list of workflows. Right now the worker is not required anymore to have the JSON defining the workflows, it loads them from the db. We can discuss how to handle this thing in a different way if you believe we can improve it. I'm not that happy about this include/exclude options at all, indeed.

> * the bigger thing is **cli task in place of async task**; I'm not excited about this because this is not how it was described in the task/epic. I know that image generation can be generalized as a sequence of cli commands, but I'm afraid at some point we'll just want a freeform async task where you can do whatever - call clis, other services, etc. I won't CR this but please let's avoid such surprises. sometimes we do change the task specs just-in-time when they are obviously wrong, but this was not strictly necessary here

My idea is a bit different here, let me explain: in Conductor you create simple workers using the conductor API to connect and listen for jobs. I imagine specialised workers in workflows be implemented as CLI commands (one, single command), in our case mender-generate-artifact or something like that.

The worker is this a binary (go, or shell script, doesn't matter from workflows point of view) you can develop and run independently from the workflows service. Workflows is just calling it with the right arguments (the ones coming from the workflows input variables). I wouldn't define a workflow containing a long list of CLI commands, just one single command doing whatever is needed to run the task.

> * nitpicks on the **TaskResult struct/API** - I feel we can do better here. instead of optional field-per-task-type, maybe just a `result` that has subtypes? in general, some polymorphic behavior would fit here I think

Ack.

**Comment by mchalski** [comment]
@tranchitella ack on the PR structure. no worries, we'll review as-is. just a rebase is in order.

the unit testing idea - I'm ok with it.

regarding include/exclude I'll take another look.

also I get what you mean by the cli task now, let's see how this goes then.

I'll be taking another look and approve or comment.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -17,6 +17,12 @@ before_script:
   - export SERVICE_IMAGE=$DOCKER_REPOSITORY:$DOCKER_TAG
   - export COMMIT_TAG="$CI_COMMIT_REF_SLUG"_"$CI_COMMIT_SHA"
 
+test:unit:
+  services:
+    - mongo:4.0
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply .gitlab-ci.yml]: Isn't Mender 2.3 going to be released with mongo 4.0?

File: api/http/router.go [hunk 4]
```diff
@@ -19,13 +19,14 @@ import (
 	"github.com/mendersoftware/workflows/store"
 )
 
+// API URL used by the HTTP router
 const (
-	ApiUrlStatus = "/status"
+	APIURLStatus = "/status"
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/router.go]: go linter in my environment suggested capitalizing it and I usually prefer to follow lint suggestions. I'm obviously okay using a different convention if we use it across the project.

**Comment by mchalski** [comment]
don't have a better idea regarding exclusion. in fact, the command line args are optional and flexible, so let's see how they work out if we already have them.

File: api/http/router.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kjaskiewiczz) [reply api/http/router.go]: on one hand, we are using camel case in similar cases everywhere
on the other hand, it's hard to discuss with golint...

File: .gitlab-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjaskiewiczz) [reply .gitlab-ci.yml]: I thought that Mender 2.3 will be using mongo 3.6
@estenberg ^^

File: api/http/router.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/router.go]: It's a warning at my side, we can live with it; don't you get the same warning in your editor? Mine is also complaining about missing comments in exported constants and functions.

File: .gitlab-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tranchitella) [reply .gitlab-ci.yml]: From what I understood, upgrading to mongo 3.6 was the first step towards mongo 4.0, which is our real target.

[the remark this answers is not in the document]
Comment (kjaskiewiczz) [reply .gitlab-ci.yml]: it was, I just don't know if we are aiming to upgrade mongo to 4.0 before the Mender 2.3 release

File: api/http/router.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kjaskiewiczz) [reply api/http/router.go]: no, my editor is not complaining, but I have a linter configurated with some high value of min_confidence and it also happens that I'm ignoring warnings about comments;
the thing with go is that it forces formatting so we do have gofmt checks, but we don't have linter policy; it's ok to introduce one if you like

**Comment by lluiscampos** [comment]
> @lluiscampos can you please review my approach to run the unit tests that need a mongodb instance?
> Commit [744f790](https://github.com/mendersoftware/workflows/commit/744f7902430967b66ed7bbb03e83cb6b19c6e2b9).

In my opinion, this is a better approach than running mongodb locally (as I have seen in other repos). I haven't reviewed in detail the tests (there are already a lot of reviewers) but the general design looks good :+1:

File: .gitlab-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjaskiewiczz) [reply .gitlab-ci.yml]: @tranchitella @estenberg do we know which version of mongo we'll be using in the mender 2.3?

[the remark this answers is not in the document]
Comment (tranchitella) [reply .gitlab-ci.yml]: @kjaskiewiczz I really hope we can release mender 2.3 with mongodb 4.0 support

**Comment by mchalski** [comment]
that would be my idea too. let's just merge and work out smaller issues as we go.
